plugins/optimization-detective/class-od-url-metric-group-collection.php

```diff
@@ -244,13 +245,13 @@ public function clear_cache(): void {
 	 * @return OD_URL_Metric_Group[] Groups.
 	 */
 	private function create_groups(): array {
-		$groups    = array();
-		$min_width = 0;
-		foreach ( $this->breakpoints as $max_width ) {
-			$groups[]  = new OD_URL_Metric_Group( array(), $min_width, $max_width, $this->sample_size, $this->freshness_ttl, $this );
-			$min_width = $max_width + 1;
+		$groups              = array();
+		$min_width_exclusive = 0;
+		foreach ( $this->breakpoints as $max_width_inclusive ) {
+			$groups[]            = new OD_URL_Metric_Group( array(), $min_width_exclusive, $max_width_inclusive, $this->sample_size, $this->freshness_ttl, $this );
+			$min_width_exclusive = $max_width_inclusive;
 		}
-		$groups[] = new OD_URL_Metric_Group( array(), $min_width, PHP_INT_MAX, $this->sample_size, $this->freshness_ttl, $this );
+		$groups[] = new OD_URL_Metric_Group( array(), $min_width_exclusive, PHP_INT_MAX, $this->sample_size, $this->freshness_ttl, $this );
```

Comment:
Key change. Note how there is no longer the need to add 1 to the maximum width to become the minimum width of the next group.

We could consider replacing `PHP_INT_MAX` here with `null`, to allow the maximum viewport width to be `positive-int|null`. The logic would then need to check if `OD_URL_Metric_Group::is_viewport_width_in_range()` the `maximum_viewport_width` is `null`, and if so it could always return `true` if the provided viewport width is greater than the `minimum_viewport_width`. This would match what I've implemented client-side in the `isViewportNeeded()` function to avoid exporting `PHP_INT_MAX` to the frontend.

Reply:
Note that `od_generate_media_query()` supports this behavior, where you can either supply `0` or `null` to omit as the minimum viewport width to omit the `min-width` media query, whereas you can also supply `null` or `PHP_INT_MAX` as the maximum viewport width to omit the `max-width`.

Reply:
I like the idea of getting rid of the need for `PHP_INT_MAX` here, to no longer have to export it, which in practice feels excessive anyway. I don't anticipate any viewports being wider than, let's say 10,000 px, and that's probably already excessive. I'm sure there are such cases, but in any case that's still miles away from `PHP_INT_MAX` :)

So no longer using a concrete number here and instead truly indicating "unbounded" makes more sense to me.

Reply:
I've implemented this in b4ea6bd66789cbdb6f6976051a9b084f01da2c92